src/main/java/build/buildfarm/cas/cfc/CASFileCache.java

```diff
@@ -2852,21 +2870,22 @@ void commit() throws IOException {
           }
         }
 
+        // FIXME: Why are we racing?
```

Comment:
To clarify we're not racing here, we raced above. If existingEntry is not null, it means that we only lost the race to create the hardlink, and now we will have an existing entry or will not reach here and have thrown for 'did not appear'
As for why we're racing: two (or more) writes that reach the closure point have an arbiter in the filesystem (why we require such strict atomicity with hard link creation for fs implementations) where they cede control and accept that the other has completed. We need to get information on completion so that we can assert a reference on valid return to have something to dereference via cleanups on the myriad usages: https://github.com/bazelbuild/bazel-buildfarm/blob/main/src/main/java/build/buildfarm/cas/cfc/CASFileCache.java#L1079, https://github.com/bazelbuild/bazel-buildfarm/blob/3319e35330efa56b2cb4aacd69f63a81eb81fe8b/src/main/java/build/buildfarm/cas/cfc/CASFileCache.java#L704. All of these things are resolutions for the state machine that is the completed write, and trying to allow other threads to slip in and perform actions on the part of already committed winners is unnecessary complication of the state machine of writes.